src/goto-programs/goto_convert_exceptions.cpp

```diff
@@ -266,39 +264,43 @@ symbol_exprt goto_convertt::exception_flag()
   return symbol_exprt(id, bool_typet());
 }
 
-void goto_convertt::unwind_destructor_stack(
+/// Unwinds the destructor stack and creates destructors for each node between
+/// destructor_start_point and destructor_end_point (including the start,
+/// excluding the end)
+bool goto_convertt::unwind_destructor_stack(
   const source_locationt &source_location,
-  std::size_t final_stack_size,
   goto_programt &dest,
-  const irep_idt &mode)
+  const irep_idt &mode,
+  optionalt<node_indext> destructor_end_point,
+  optionalt<node_indext> destructor_start_point)
 {
-  unwind_destructor_stack(
-    source_location, final_stack_size, dest, targets.destructor_stack, mode);
+  return unwind_destructor_stack(
+    source_location,
+    dest,
+    targets.destructor_stack,
+    mode,
+    std::move(destructor_end_point),
+    std::move(destructor_start_point));
 }
 
-void goto_convertt::unwind_destructor_stack(
+bool goto_convertt::unwind_destructor_stack(
   const source_locationt &source_location,
-  std::size_t final_stack_size,
   goto_programt &dest,
-  destructor_stackt &destructor_stack,
-  const irep_idt &mode)
+  destructor_treet &destructor_stack,
+  const irep_idt &mode,
+  optionalt<node_indext> destructor_end_point,
+  optionalt<node_indext> destructor_start_point)
 {
-  // There might be exceptions happening in the exception
-  // handler. We thus pop off the stack, and then later
-  // one restore the original stack.
-  destructor_stackt old_stack=destructor_stack;
+  const std::vector<codet> stack = destructor_stack.get_destructors(
+    std::move(destructor_end_point), std::move(destructor_start_point));
 
-  while(destructor_stack.size()>final_stack_size)
+  for(const codet &destructor : stack)
```

Comment:
I think we need to continue to do what the code on the left says -- i.e. adjust the stack position before we execute each destructor, so that e.g. if the destructor itself throws, then we don't re-execute destructors. For example, the destructor might be `{ int x; throw 0; }`, in which case that `throw 0;` should kill `x` and any variables on the current stack that we haven't yet destroyed, but should not re-destroy things on the stack that we've already killed.